llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -2960,6 +2969,204 @@ static bool validateAndCostRequiredSelects(BasicBlock *BB, BasicBlock *ThenBB,
+  if (!HoistLoadsStoresWithCondFaulting ||
+      !TTI.hasConditionalLoadStoreForType())
+    return false;
+
+  if (!BI || !BI->isConditional())
+    return false;
+
+  SmallVector<BasicBlock *, 2> Successors;
+  if (ThenBB)
+    Successors.push_back(ThenBB);
+  else
+    Successors.append({BI->getSuccessor(0), BI->getSuccessor(1)});
+
+  // If either of the blocks has it's address taken, then we can't do this fold,
+  // because the code we'd hoist would no longer run when we jump into the block
+  // by it's address.
+  for (auto *Succ : Successors)
+    if (Succ->hasAddressTaken())
+      return false;
+
+  // Not use isa<AllocaInst>(getUnderlyingObject(I.getOperand(0)) to avoid
+  // checking all intermediate operands dominate the branch.
+  auto IsLoadFromAlloca = [](const Instruction &I) {
+    return isa<LoadInst>(I) && isa<AllocaInst>((I.getOperand(0)));
+  };
+
+  // Collect hoisted loads/stores.
+  SmallSetVector<Instruction *, 4> HoistedInsts;
+  // Not hoist load/store if
+  // 1. target does not have corresponding conditional faulting load/store.
+  // 2. it's volatile or atomic.
+  // 3. there is a load/store that can not be hoisted in the same bb.
+  // 4. there is a non-load/store that's not safe to speculatively execute
+  //    in the same bb.
+  // 5. any operand of it does not dominate the branch.
+  // 6. it's a store and a memory read is skipped.
+  auto HoistInstsInBB = [&](BasicBlock *BB) {
+    bool SkipMemoryRead = false;
+    // A more efficient way to check domination. An operand dominates the
+    // BranchInst if
+    // 1. it's not defined in the same bb as the instruction.
+    // 2. it's to be hoisted.
+    //
+    // b/c BB is only predecessor and BranchInst does not define any value.
+    auto OpsDominatesBranch = [&](Instruction &I) {
+      return llvm::all_of(I.operands(), [&](Value *Op) {
+        if (auto *J = dyn_cast<Instruction>(Op)) {
+          if (HoistedInsts.contains(J))
+            return true;
+          if (J->getParent() == I.getParent())
+            return false;
+        }
+        return true;
+      });
+    };
+    for (auto &I : *BB) {
+      auto *LI = dyn_cast<LoadInst>(&I);
+      auto *SI = dyn_cast<StoreInst>(&I);
+      if (LI || SI) {
+        bool IsSimple = (LI && LI->isSimple()) || (SI && SI->isSimple());
+        if (!IsSimple || !OpsDominatesBranch(I))
+          return false;
+        auto *Type = getLoadStoreType(&I);
+        // a load from alloca is always safe.
+        if (!IsLoadFromAlloca(I) && !TTI.hasConditionalLoadStoreForType(Type))
+          return false;
+        // Conservative aliasing check.
+        if (SI && SkipMemoryRead)
+          return false;
+        HoistedInsts.insert(&I);
+      } else if (!I.isTerminator() && !isSafeToSpeculativelyExecute(&I))
+        return false;
+      else if (I.mayReadFromMemory())
+        SkipMemoryRead = true;
+    }
+    return true;
+  };
+
+  for (auto *Succ : Successors)
+    if (!HoistInstsInBB(Succ))
```

Comment:
Missed optimization:
```
define void @test(ptr %p, i1 %mask, i32 %val) {
entry:
  br i1 %mask, label %if.then, label %if.else

if.then:
  store i32 %val, ptr %p, align 4
  br label %exit

if.else:
  call void @sideeffect()
  br label %exit

exit:
  ret void
}

declare void @sideeffect()
```

Reply:
@dtcxzyw call intrinsic can only be hoisted only when it has attribute `speculatable`.

Reply:
> @dtcxzyw call intrinsic can only be hoisted only when it has attribute `speculatable`.

I mean we can hoist the store instruction and keep `%if.else` as is.
```
define void @test(ptr %p, i1 %mask, i32 %val) {
entry:
  ; masked store i32 %val, ptr %p, align 4
  br i1 %mask, label %if.then, label %if.else

if.then:
  br label %exit

if.else:
  call void @sideeffect()
  br label %exit

exit:
  ret void
}

declare void @sideeffect()
```

Reply:
Is it valuable to hoist if we aren't eliminating branches?

Reply:
In this case we can further eliminate the edge `%if.then -> %exit`:
```
define void @test(ptr %p, i1 %mask, i32 %val) {
entry:
  ; masked store i32 %val, ptr %p, align 4
  br i1 %mask, label %exit, label %if.else

if.else:
  call void @sideeffect()
  br label %exit

exit:
  ret void
}

declare void @sideeffect()
```

Reply:
Hmm, this seems roughly on the edge of profitability. If the `exit` block couldn't be folded into the `if.then`/`if.else` there would be a better case though. At least for my money, the initial patch should only include cases where we can fully eliminate the if/else. That case is more clearly profitable and think it makes sense to limit the initial implementation (when we are operating with no performance data) in such a way.